### ISSUE TITLE
jupyterlmod does not work with Jupyter Lab

### ISSUE THREAD

**Issue by dr-br** [cmd-ntrf/jupyter-lmod#26]
~No Software is listed in the "Loaded Modules" or "Available Modules" sections.~
~When I load modules in notebook (version 1.9.0, see #25 ), they are not displayed neither.~

~I tried different Jupyter {lab, notebook, ...} versions and many jupyterlmod versions and some jupyterlab-lmod@XXX versions. No success.~

The most recent version of jupyterlmod which works on our cluster is:
jupyterlmod==1.9.0 (pip)
jupyterlab-lmod@0.7.0 (labextension)

```
jupyter --version
jupyter core     : 4.6.3
jupyter-notebook : 6.1.4
qtconsole        : 4.7.7
ipython          : 7.16.1
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 2.2.8
nbconvert        : 6.0.5
ipywidgets       : 7.5.1
nbformat         : 5.0.7
traitlets        : 4.3.3

pip freeze|grep jupyter
jupyter==1.0.0
jupyter-client==6.1.7
jupyter-console==6.2.0
jupyter-core==4.6.3
jupyter-server-proxy==1.5.0
jupyterlab==2.2.8
jupyterlab-pygments==0.1.1
jupyterlab-server==1.2.0
jupyterlmod==1.9.0

jupyter labextension list
JupyterLab v2.2.8
Known labextensions:
   app dir: /pfs/data5/home/kit/scc/ej4555/lmodtestenv/lmodtestenv/share/jupyter/lab
        jupyterlab-lmod v0.8.0  enabled  OK

module --version

Modules based on Lua: Version 8.3.15  2020-06-05 12:32 -05:00
    by Robert McLay mclay@tacc.utexas.edu
```

It would be fantastic, if the newest versions of jupyterlmod (python and npm package) would work on our cluster.
Thanks!

**Comment by dr-br** [comment]
Would you please provide any hint, how to debug in order to give you useful information?
I updated all software, the current versions are now: 
```
jupyter --version
jupyter core     : 4.7.0
jupyter-notebook : 6.1.5
qtconsole        : 5.0.0
ipython          : 7.16.1
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 2.2.9
nbconvert        : 6.0.7
ipywidgets       : 7.5.1
nbformat         : 5.0.8
traitlets        : 4.3.3

pip freeze|grep jupyter
jupyter-client==6.1.7
jupyter-core==4.7.0
jupyter-telemetry==0.1.0
jupyterhub==1.2.1
jupyterlab==2.2.9
jupyterlab-pygments==0.1.2
jupyterlab-server==1.2.0
jupyterlmod==2.0.1

jupyter labextension list
JupyterLab v2.2.9
Known labextensions:
   app dir: /pfs/data4/software_uc2/bwhpc/common/devel/jupyter_base/share/jupyter/lab
        jupyterlab-lmod v0.8.0  enabled  OK

module --version

Modules based on Lua: Version 8.3.15  2020-06-05 12:32 -05:00
    by Robert McLay mclay@tacc.utexas.edu

node --version
v10.21.0
```
It still does not work. Neither classic notebook nor JupyterLab. The old combination 
jupyterlmod==1.9.0 (pip)
jupyterlab-lmod@0.7.0 (labextension)
works.

**Comment by dr-br** [comment]
Maybe Browser output helps:
![Bildschirmfoto von 2020-11-24 15-34-58](https://user-images.githubusercontent.com/45423137/100108215-d4691680-2e6a-11eb-921f-498a71a2a5bf.png)

**Comment by cmd-ntrf** [comment]
The screenshot is a good start.

Can you expand the post request that returns a 400 error?

**Comment by dr-br** [comment]
which one?
There are several tabs. Haders, Cookies, ...

**Comment by cmd-ntrf** [comment]
I meant this :
![image](https://user-images.githubusercontent.com/884095/100113804-b8e81680-2e3e-11eb-8f56-42d8ee8d8107.png)

**Comment by dr-br** [comment]
![Bildschirmfoto von 2020-11-24 16-21-56](https://user-images.githubusercontent.com/45423137/100114204-77bd2a00-2e71-11eb-99dc-b064f65e2de0.png)

**Comment by dr-br** [comment]
![Bildschirmfoto von 2020-11-24 16-21-56](https://user-images.githubusercontent.com/45423137/100114319-94f1f880-2e71-11eb-9498-3368e492111a.png)

**Comment by dr-br** [comment]
any other tab which might be interesting?

**Comment by dr-br** [comment]
maybe chromium is more helpful
![Bildschirmfoto von 2020-11-24 16-28-47](https://user-images.githubusercontent.com/45423137/100114935-3c6f2b00-2e72-11eb-97da-7e817505d175.png)

**Comment by cmd-ntrf** [comment]
Can you try to locate the JSON that was posted by the request?

The 400 comes from this part of the jupyterlmod handler:
https://github.com/cmd-ntrf/jupyter-lmod/blob/master/jupyterlmod/handler.py#L38

The posted payload does not match what is expected, so the handler returns a 400. The javascript part of jupyter-lmod is what is posting JSON data to the jupyter-lmod handler, so the format should match, but we would know for sure by looking at the json sent to the handler by your browser.

**Comment by dr-br** [comment]
> Can you try to locate the JSON that was posted by the request?

how?

**Comment by cmd-ntrf** [comment]
This will not provide the json, but can you click on this  (lmod.js) :
![image](https://user-images.githubusercontent.com/884095/100115811-f8176700-2e40-11eb-8f86-f5b97fce40a1.png)

And provide a screenshot of the lines surrounding line 34?

**Comment by dr-br** [comment]
![Uploading Bildschirmfoto von 2020-11-24 16-40-10.png…]()

**Comment by dr-br** [comment]
![Bildschirmfoto von 2020-11-24 16-40-10](https://user-images.githubusercontent.com/45423137/100116429-059a1480-2e74-11eb-90a6-9d2e3d1efc30.png)

**Comment by cmd-ntrf** [comment]
The jupyter notebook logs should include the 400 error and tell us which one of the two 400 errors it is  (either modules is absent from the json, or modules is not a list).

To retrieve the json from Chrome, there are instructions on stackoverflow:
https://stackoverflow.com/questions/15603561/how-can-i-debug-a-http-post-in-chrome

**Comment by dr-br** [comment]
like i already wrote: module contains `[]`

**Comment by cmd-ntrf** [comment]
modules is the list of available modules, not the json sent with the post.

That said, it should not be empty. If it is empty, it means the handler or the communication with lmod is not working properly. I don't know why at this point. I will have to look at what change in that regard between 1.9 and 2.0.

**Comment by dr-br** [comment]
![Bildschirmfoto von 2020-11-24 17-04-43](https://user-images.githubusercontent.com/45423137/100119541-434c6c80-2e77-11eb-9209-43c9a6263d82.png)

**Comment by dr-br** [comment]
"Remove jquery from jupyterlab requirements" on July, 2
Could the first error (tooltip is not a function) be associated to this?

**Comment by cmd-ntrf** [comment]
The problem appears to come from the lmod Python module that does not return the list of available modules.

Since version 2.0 of jupyter-lmod, the parsing of lmod output is done with regular expressions : 
https://github.com/cmd-ntrf/jupyter-lmod/blob/master/lmod/__init__.py#L15

It is possible that your version of lmod output mismatch the regular expressions and jupyter-lmod fails in parsing lmod output. To test this, install jupyterlmod in a virtual environment on your cluster, and in an ipython shell execute the following commands:
```python
%autoawait
import lmod
await lmod.avail()
```

It should return the list of available modules. If this list is empty, there is a problem with the regex or lmod output. If it is not empty, the problem lies elsewhere and I'll keep digging.

**Comment by dr-br** [comment]
Hi, nice debug hint!
```
python -m venv testlmod
source testlmod/bin/activate
pip install --upgrade pip
pip install git+https://github.com/cmd-ntrf/jupyter-lmod

ipython

Python 3.6.8 (default, Dec  5 2019, 15:45:45) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.
In [1]: %autoawait
IPython autoawait is `on`, and set to use `asyncio`
In [2]: import lmod
In [3]: await lmod.avail()
Out[3]: []
In [4]: 
```
Is there a way to debug further? Strace doesn't seem to give meaningful insight...

**Comment by dr-br** [comment]
Hi,
we found the error:
Our modules are arranged with depth 3 instead of 2.
The original regex `MODULE_REGEX` assumes a depth of 2.
So `devel/myModule1.2.3` is found, `devel/myModule/version1.2.3` is not.
The regex below handles a depth of 2 and 3.
Furthermore, we (actually, my colleague who knows regex) think, that `…A-Z…` is more appropriate than `…A-z…`
```
MODULE_REGEX = re.compile(r"^([a-zA-Z0-9-_+.]+[/][a-zA-Z0-9-_+.]+|[a-zA-Z0-9-_+.]+[/][a-zA-Z0-9-_+.]+[/][a-zA-Z0-9-_+.]+)$", re.M)
```
We do not propose any modification for `MODULE_REGEX_NO_HIDDEN` because we didn't get its meaning.

Sorry, this is unavoidable:
https://xkcd.com/208/

Thank you very much @cmd-ntrf for your help!

**Comment by cmd-ntrf** [comment]
I am glad we could find the issue.

The original regex can be simplified to work with any hierarchy depth:
```
^[a-zA-Z0-9-_+.\/]{1,}[^\/:]$
```
You can try it here : https://regex101.com/r/003wfS/1

`MODULE_REGEX_NO_HIDDEN`, is the mostly same regex, but its purpose is to remove the modules that are hidden in lmod. Hidden modules have dot (`.`) at the beginning of their version number for example `devel/gcc/.7.3.0` would be hidden.

This regex will remove the hidden module and work with any depth:
```
^[a-zA-Z0-9-_+.\/]{1,}\/[a-zA-Z0-9][a-zA-z0-9-_.]*[^\/:]$
```

I will test this thoroughly today and release jupyter-lmod 2.0.2.

Thank you for your patience and help in debugging this.

**Comment by cmd-ntrf** [comment]
Slightly improve version of `MODULE_REGEX_NO_HIDDEN` to allow modules with no slashes:
```
^([^.][[a-zA-Z0-9-_+.]*|[a-zA-Z0-9-_+.\/]{1,}\/[a-zA-Z0-9][a-zA-z0-9-_.]*[^\/:])$
```

**Comment by cmd-ntrf** [comment]
@dr-br : can you try branch `fix_regex` and confirm it fixes your issue?

Thanks

**Comment by dr-br** [comment]
Works.
Also resolves #25 !
Tnaks a lot!
Samuel
